### DISCUSSION TITLE
Should I be increasing the connection limits or doing something else?

### DISCUSSION THREAD

**Discussion by huntfx** [sqlalchemy/sqlalchemy#6077] [Usage Questions] [answered]
I've just launched a script at work (using MariaDB), and so far its limited to people with usernames beginning with "a". I noticed MySQL workbench reporting 50+ connections open at the same time (out of a limit of 150). 

According to [this question](https://stackoverflow.com/q/30036597), it's expected behaviour, even though I'm careful with opening/closing sessions (each session lasts less than a second), and 99% of the time I'm not actually opening any sessions.

I'm a bit worried about potential errors if I was to set it to run for more people, so what would be the best practice in this case? 
Should I be setting a very small `pool_timeout`, manually attempting to invalidating sessions, asking IT to increase the connection limit, or just relying on the database itself to forget old connections when a new one is made?

**Attempts**
I set `pool_timeout=5` in `create_engine`, and the connection never actually dropped while the application was open. Using `echo_pool='debug'` showed it kept reusing the same connection object even after I left it alone for way longer than 5 seconds.

Using `session.invalidate()` instead of `session.close()` did actually stop the issue, but obviously that adds a lot of overhead. If I went down that route, I'd have to only run the invalidation on certain events, which would not be easily maintainable.

**Edit**
Just saw the `wait_timeout` option in the database, but apparently changing it [can cause issues](https://stackoverflow.com/q/21076306)?

**Comment by zzzeek** [comment, marked as answer]
> I've just launched a script at work (using MariaDB), and so far its limited to people with usernames beginning with "a". I noticed MySQL workbench reporting 50+ connections open at the same time (out of a limit of 150).

So top level thing about MariaDB  / MySQL, 150 is really low.    At Red Hat with Openstack we regularly swim in the area of ~2500 connections, the vast majority of them idle, and some large envs can get as high as 8K.      Now, that is not a **good** thing, this has to do with the fact that Openstack spins up a huge amount of processes each with these connection pools in it, but the idea is that, **mariadb has no problem with it**.   You can have hundreds of those connections doing things at once and nothing bad happens.   you just want to make sure you set innodb_buffer_pool_size to be at least 1G or more.

> 
> According to [this question](https://stackoverflow.com/q/30036597), it's expected behaviour, even though I'm careful with opening/closing sessions (each session lasts less than a second), and 99% of the time I'm not actually opening any sessions.

the "mysql server has gone away" can happen for more than one reason but the "normal" one is the MySQL idle timeout.   so you just set up pool_recycle=3600 or something like that and you won't see that particular issue anymore.

> 
> I'm a bit worried about potential errors if I was to set it to run for more people, so what would be the best practice in this case?
> Should I be setting a very small `pool_timeout`, manually attempting to invalidating sessions, asking IT to increase the connection limit, or just relying on the database itself to forget old connections when a new one is made?

pool_timeout is usually fine at its default of 30 seconds but can be lowered if you want an exhausted connection pool to fail quickly, rather than keeping people waiting.  this is not the same thing as pool_recycle.   im not sure about the "manually invalidating sessions" part, do you mean MySQL sessions?   i dont think any special action is needed there.    

There is a MySQL pooling middleware called [proxysql](https://proxysql.com/) that I think is somewhat popular, I didnt find the documentation to be that good but it's been around a long time.    if you wanted to play with that, you'd set up NullPool in your SQLAlchemy applications and then proxysql would, we hope, do a more coordinated job of pooling across lots of processes, if that's the kind of thing you are going for.


> 
> **Attempts**
> I set `pool_timeout=5` in `create_engine`, and the connection never actually dropped while the application was open. Using `echo_pool='debug'` showed it kept reusing the same connection object even after I left it alone for way longer than 5 seconds.

yeah that's not what pool_timeout does.  if you want to see the  5 seconds, set the pool size to 5 and the max overflow to 0.  then check out 5 connections.   then when you check out the 6th one, it will block for 5 seconds, then raise an error.  this timeout is how long it waits to give someone a connection when there are no connections available.   you get this error: https://docs.sqlalchemy.org/en/14/errors.html#queuepool-limit-of-size-x-overflow-y-reached-connection-timed-out-timeout-z 

> 
> Using `session.invalidate()` instead of `session.close()` did actually stop the issue, but obviously that adds a lot of overhead. If I went down that route, I'd have to only run the invalidation on certain events, which would not be easily maintainable.

i didnt even realize there was a session.invalidate() method.   this looks like it was added for edge scenarios like gevent timeouts.    calling invalidate() on connections is not something you would normally want to do, it just evicts the connection from the pool but there's no reason to do that unless something has gone very wrong with the connection.

> 
> **Edit**
> Just saw the `wait_timeout` option in the database, but apparently changing it [can cause issues](https://stackoverflow.com/q/21076306)?


yes this is MySQL/ MariaDB's idle connection timeout.  you want your pool_recycle number of seconds to be lower than the databases's wait_timeout.

you probably also want to look into using the pre_ping option which is very useful to detect liveness on a connection each time you go to use it.

**Comment by ehdgua01** [comment]
Hi, @zzzeek 

When I use `pool_pre_ping` instead of `pool_recycle`?
Which of the two options is better?
I can solve just using `pool_pre_ping` but I assumed that `pool_pre_ping` has an overhead because that checks the connection for each request. Is that right?

**Reply by zzzeek** [reply]
you should use both IMO.   pool_recycle will jump in early on connections that are stale and pool_pre_ping will make sure everything that's not known to be stale is definitely fresh.     pre_ping's overhead is low in relation to how frequent a "checkout" is in modern SQLAlchemy.   it's more expensive in the case that it actually sees it has to reconnect though, so pool_recycle will eliminate that problem for old connections (assuming your DB closes stale connections).
